tests/mock_network/messages.rs

```diff
@@ -73,69 +63,58 @@ fn send_and_receive() {
     let mut nodes = create_connected_nodes(&env, elder_size + 1);
 
     let sender_index = gen_elder_index(&mut rng, &nodes);
-    let src = SrcLocation::Node(*nodes[sender_index].id());
-    let dst = DstLocation::Section(rng.gen());
-
+    let req_src = SrcLocation::Node(*nodes[sender_index].id());
+    let req_dst = DstLocation::Section(rng.gen());
     let req_content = gen_vec(&mut rng, 10);
     let res_content = gen_vec(&mut rng, 11);
 
     assert!(nodes[sender_index]
         .inner
-        .send_message(src, dst, req_content.clone())
+        .send_message(req_src, req_dst, req_content.clone())
         .is_ok());
 
-    let _ = poll_all(&mut nodes);
-
-    let mut request_received_count = 0;
-
-    for node in nodes
-        .iter_mut()
-        .filter(|n| n.inner.is_elder() && n.in_dst_location(&dst))
-    {
-        loop {
-            match node.try_recv_event() {
-                Some(Event::MessageReceived { content, src, .. }) => {
-                    request_received_count += 1;
-                    if req_content == content {
-                        let res_src = SrcLocation::Section(*node.our_prefix());
-                        let res_dst = match src {
-                            SrcLocation::Node(id) => DstLocation::Node(*id.name()),
-                            _ => panic!("Unexpected src location: {:?}", src),
-                        };
-
-                        if let Err(err) =
-                            node.inner
-                                .send_message(res_src, res_dst, res_content.clone())
-                        {
-                            trace!("Failed to send message: {:?}", err);
-                        }
-                        break;
-                    }
-                }
-                Some(_) => (),
-                _ => panic!("Event::MessageReceived not received"),
-            }
-        }
-    }
-
-    assert!(request_received_count >= quorum);
+    // For all expected recipients, poll until it receives the request message and send
+    // the response
+    let expected_req_recipients: Vec<_> = expected_recipients(&nodes, &req_dst).collect();
+    assert!(expected_req_recipients.len() >= quorum);
 
-    let _ = poll_all(&mut nodes);
+    for index in expected_req_recipients {
+        poll_until(&env, &mut nodes, |nodes| {
```

Comment:
I think when execute the first iteration of poll_until `index = 0`, there will be high chance that every nodes got request received and responded.
Though it shall be no-harm for the later on iterations, it feels bit un-efficient?

Reply:
`poll_until` evaluates the termination predicate as the very first thing, so if it's already true, it won't run even single iteration. So efficiency-wise, this should be fine.